### DISCUSSION TITLE
Is it possible to debug code when using Mill?

### DISCUSSION THREAD

**Discussion by hmf** [com-lihaoyi/mill#1076] [General]
I am wondering if there is any way we can start execution in debug mode. More specifically is it possible using an IDE such as IJ to step over the code?

Maybe we have a `-jvm-debug` flag [like SBT](https://stackoverflow.com/questions/4150776/debugging-scala-code-with-simple-build-tool-sbt-and-intellij)?

TIA

**Comment by lefou** [comment]
Do you want to debug mill itself or the source code of your module (located under `src`)?

**Reply by hmf** [reply]
Debug the source code of the project located under `src` and `test`.

**Reply by Shri333** [reply]
Is it possible to debug mill itself? I tried attaching a remote debugger to the root module but it did not work.

**Comment by lefou** [comment]
You can control how your code is executed, e.g. when you use `run` or `test`, mill will respect the targets `forkArgs`, `forkEnv` and `forkWorkingDir` when starting a dedicated JVM. 

You could enable remote debugging then:

```scala
override def forkArgs = Seq("-agentlib:jdwp=transport=dt_socket,server=y,address=8000,suspend=y")
```

The JVM will wait for a remote debugger to connect to port 8000 before it starts to process your program or test.

I'm pretty sure that the BSP protocol also allows you to run the program directly from the IDE, but I never have tried it.

**Reply by hmf** [reply]
Ok, I will try this with IJ. Link I gave already has IJ set-up. Thank you.

**Reply by lefou** [reply]
Did it work for you?

**Reply by hmf** [reply]
Did not try it. I switched to VSCode + Metals plugin (uses Bloop). It automatically places a run/debug position over the main functions. All I need to do is click on "debug" and I am good to go.

**Comment by Sailsman63** [comment]
I know that this is an old topic, but for the curious:

I've been playing around with Intellij support a bit, and have found the following:

- BSP *may* have the option to run programs (and arbitrary build-targets?), but Intellij Idea Scala plugin's support for it stinks. It only does incremental builds (`__.compile` on file change) and test runs. (`mill __.test` on demand.) You cannot attach a debugger to the test run.
- You can run arbitrary targets as "Shell Script" build configurations. `forkArgs` etc. are respected (as expected from command line mill), but you need to manually connect to the new process from the run menu.
- Best debug integration that I've managed so far is to have two build configurations:
    - A "Shell Script" configuration that triggers a mill task that deploys a live jar locally
    - A "Jar Application" configuration that points at the jar deployed above. This task can automatically attach the debugger. Put the first task in the "Before Launch" list for this task, to make sure you're always using the up-to-date code.

EDIT: and of course, there turns out to be a caveat. As of this time, Windows users will be unable to do the above, due to https://youtrack.jetbrains.com/issue/IDEA-277486 (Unclear how simple the workaround is to implement)

**Reply by hmf** [reply]
My experience with Scala 3 + IJ has not been good either. It seems that only when you explicitly create a Scala 3 project via the menu do you get the support.
